cilium-cli: extend no-interrupted-connections to test Egress Gateway

ysksuzuki opened the pull request:
This PR extends the conn-disrupt-test to ensure that connections through a Egress Gateway are not disrupted.

The test deploys a conn-disrupt-test server with a single replica on an external node, a client on a gateway node and another client on a non gateway node. Clients establish long-lived TCP connections with the server through the gateway node.

After upgrade, the test checks the restart counters, and compares them with the counters stored before the upgrade. A mismatch indicates that a connection was interrupted.

Example output from `kubectl -n cilium-test-1 get po -o wide`:
```
NAME                                                        READY   STATUS    RESTARTS      AGE   IP             NODE                 NOMINATED NODE   READINESS GATES
test-conn-disrupt-client-egw-gw-node-859d6b6875-vkvh2       1/1     Running   0             15s   10.244.3.206   kind-worker          <none>           <none>
test-conn-disrupt-client-egw-non-gw-node-6c69d885df-7z6c5   1/1     Running   0             15s   10.244.1.14    kind-worker2         <none>           <none>
test-conn-disrupt-server-egw-644f65cb-vz8v7                 1/1     Running   0             17s   172.18.0.4     kind-worker3         <none>           <none>
```
Fixes: #37092

ysksuzuki committed:
cilium-cli: extend no-interrupted-connections to test Egress Gateway

This commit extends the conn-disrupt-test to ensure that connections
through a Egress Gateway are not disrupted.

The test deploys a conn-disrupt-test server with a single replica
on an external node, a client on a gateway node and another client
on a non gateway node. Clients establish long-lived TCP connections
with the server through the gateway node.

After upgrade, the test checks the restart counters, and compares them
with the counters stored before the upgrade. A mismatch indicates that
a connection was interrupted.

Example output from `kubectl -n cilium-test-1 get po -o wide`:
```
NAME                                                        READY   STATUS    RESTARTS      AGE   IP             NODE                 NOMINATED NODE   READINESS GATES
test-conn-disrupt-client-egw-gw-node-859d6b6875-vkvh2       1/1     Running   0             15s   10.244.3.206   kind-worker          <none>           <none>
test-conn-disrupt-client-egw-non-gw-node-6c69d885df-7z6c5   1/1     Running   0             15s   10.244.1.14    kind-worker2         <none>           <none>
test-conn-disrupt-server-egw-644f65cb-vz8v7                 1/1     Running   0             17s   172.18.0.4     kind-worker3         <none>           <none>
```

Signed-off-by: Yusuke Suzuki <yusuke.suzuki@isovalent.com>

ysksuzuki committed:
ci: enable conn-disrupt-test for Egress Gateway

This commit enables conn-disrupt-test for Egress Gateway.

Signed-off-by: Yusuke Suzuki <yusuke.suzuki@isovalent.com>

ysksuzuki committed:
cilium-cli: skip conn-disrupt-test EGW when lb-acceleration is enabled

Restarting the agent disrupts the connection through the EGW
when lb-acceleration is enabled due to #37431
Let's skip it for now

Signed-off-by: Yusuke Suzuki <yusuke.suzuki@isovalent.com>

giorio94 commented on cilium-cli/connectivity/check/wait.go:
```diff
@@ -343,6 +343,96 @@ func WaitForIPCache(ctx context.Context, log Logger, agent Pod, pods []Pod) erro
+		nextTargetEntry:
+			for _, targetEntry := range targetEntries {
+				for _, entry := range entries {
+					if targetEntry.matches(entry) {
```
Nit. `!slices.ContainsFunc` is slightly more idiomatic (same below). Another possibility could be sorting the two arrays and then comparing them.

giorio94 commented on cilium-cli/connectivity/check/deployment.go:
```diff
@@ -581,6 +668,51 @@ func (ct *ConnectivityTest) deploy(ctx context.Context) error {
+				return fmt.Errorf("unable to create CiliumEgressGatewayPolicy %s: %w", cegp.Name, err)
+			}
+
+			if err := ct.createTestConnDisruptServerDeployAndSvc(ctx, testConnDisruptServerEgressGatewayDeploymentName, KindTestConnDisruptEgressGateway, 1,
+				testConnDisruptEgressGatewayServiceName, testConnDisruptServerEgressGatewayAppLabel, true, newConnDisruptCNPForEgressGateway); err != nil {
+				return err
+			}
+
+			if err := ct.createTestConnDisruptClientDeployment(ctx, testConnDisruptClientEgressGatewayOnGatewayNodeDeploymentName, KindTestConnDisruptEgressGateway,
+				testConnDisruptClientEgressGatewayOnGatewayNodeAppLabel, fmt.Sprintf("test-conn-disrupt-egw.%s.svc.cluster.local.:8000", ct.params.TestNamespace),
+				1, false, map[string]string{"kubernetes.io/hostname": gatewayNode}); err != nil {
+				return err
+			}
+			if err := ct.createTestConnDisruptClientDeployment(ctx, testConnDisruptClientEgressGatewayOnNonGatewayNodeDeploymentName, KindTestConnDisruptEgressGateway,
+				testConnDisruptClientEgressGatewayOnNonGatewayNodeAppLabel, fmt.Sprintf("test-conn-disrupt-egw.%s.svc.cluster.local.:8000", ct.params.TestNamespace),
+				1, false, map[string]string{"kubernetes.io/hostname": nonGatewayNode}); err != nil {
+				return err
+			}
```
This will not work in the context of clustermesh, because it would create the server in one cluster and the clients in the other, but cross-cluster EGW policies are not supported at the moment. Either the test should be disabled when testing against clustermesh, or client and server should be always forced to the same cluster.

ysksuzuki replied:
I will disable the test when testing against clustermash.

giorio94 commented on .github/workflows/tests-clustermesh-upgrade.yaml:
```diff
@@ -463,7 +463,10 @@ jobs:
+            --include-unsafe-tests \
+            --include-conn-disrupt-test \
+            --include-conn-disrupt-test-ns-traffic \
+            --include-conn-disrupt-test-egw \
```
EGW is not enabled in this test (and doesn't work cross-cluster) so this is not strictly necessary here.

giorio94 commented on .github/workflows/tests-clustermesh-upgrade.yaml:
```diff
@@ -463,7 +463,10 @@ jobs:
-            --include-conn-disrupt-test --include-conn-disrupt-test-ns-traffic \
+            --include-unsafe-tests \
+            --include-conn-disrupt-test \
+            --include-conn-disrupt-test-ns-traffic \
```
For some reason this test does not seem to be run: https://github.com/cilium/cilium/actions/runs/13890711080/job/38862133096#step:27:255

ysksuzuki replied:
Looks like that's because NodeWithoutCilium isn't enabled. I will drop it.

giorio94 replied:
Ah fair enough, I got confused with the conformance clustermesh tests that instead set `--nodes-without-cilium`. I think it is worthy enabling the N/S disruption test in this workflow as well, but we can leave that for a follow-up PR, given that it is unrelated from this one (and so that it is easy to troubleshoot possible problems).

giorio94 reviewed (commented):
Thanks! A couple of minor comments inline.

ysksuzuki committed:
cilium-cli: skip conn-disrupt-test EGW when testing against clustermesh

Currently, the cross-cluster EGW policies are not supported.
Let's skip it for now.

Signed-off-by: Yusuke Suzuki <yusuke.suzuki@isovalent.com>

pippolo84 reviewed (approved):
Nice! :rocket:
